lib/integrations/awsoidc/eks_enroll_clusters.go

```diff
@@ -0,0 +1,558 @@
+			if err != nil {
+				log.WithError(err).Debugf("failed to enroll EKS cluster %q", eksClusterName)
+			}
+
+			mu.Lock()
+			defer mu.Unlock()
+			results = append(results, EnrollEKSClusterResult{ClusterName: eksClusterName, ResourceId: resourceId, Error: trace.Wrap(err)})
+
+			return nil
+		})
+	}
+	// We don't return error from individual group goroutines, they are gathered in the returned value.
+	_ = group.Wait()
+
+	return &EnrollEKSClusterResponse{Results: results}
+}
+
+func enrollEKSCluster(ctx context.Context, log logrus.FieldLogger, clock clockwork.Clock, credsProvider aws.CredentialsProvider, clt EnrollEKSCLusterClient, proxyAddr, clusterName string, req EnrollEKSClustersRequest) (string, error) {
+	eksClusterInfo, err := clt.DescribeCluster(ctx, &eks.DescribeClusterInput{
+		Name: aws.String(clusterName),
+	})
+	if err != nil {
+		return "", trace.Wrap(err, "unable to describe EKS cluster")
+	}
+	eksCluster := eksClusterInfo.Cluster
+
+	if eksCluster.Status != eksTypes.ClusterStatusActive {
+		return "", trace.BadParameter(`can't enroll EKS cluster %q - expected "ACTIVE" state, got %q.`, clusterName, eksCluster.Status)
+	}
+
+	principalArn, err := getAccessEntryPrincipalArn(ctx, clt.GetCallerIdentity)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	wasAdded, err := maybeAddAccessEntry(ctx, clusterName, principalArn, clt)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if wasAdded {
+		// If we added access entry, we'll clean it up when function stops executing.
+		defer func() {
+			if _, err := clt.DeleteAccessEntry(ctx, &eks.DeleteAccessEntryInput{
+				ClusterName:  aws.String(clusterName),
+				PrincipalArn: aws.String(principalArn),
+			}); err != nil {
+				log.WithError(err).Warnf("could not delete access entry for principal %q on cluster %q", principalArn, clusterName)
+			}
+		}()
+	}
+
+	_, err = clt.AssociateAccessPolicy(ctx, &eks.AssociateAccessPolicyInput{
+		AccessScope: &eksTypes.AccessScope{
+			Namespaces: nil,
+			Type:       eksTypes.AccessScopeTypeCluster,
+		},
+		ClusterName:  aws.String(clusterName),
+		PolicyArn:    aws.String(eksClusterAdminPolicy),
+		PrincipalArn: aws.String(principalArn),
+	})
+	if err != nil {
+		return "", trace.Wrap(err, "unable to associate EKS Access Policy to cluster %q", clusterName)
+	}
+
+	kubeClientGetter, err := getKubeClientGetter(ctx, clock.Now(), credsProvider, clusterName, req.Region,
+		aws.ToString(eksCluster.CertificateAuthority.Data), aws.ToString(eksCluster.Endpoint))
+	if err != nil {
+		return "", trace.Wrap(err, "unable to build kubernetes client for EKS cluster %q", clusterName)
+	}
+
+	if alreadyInstalled, err := clt.CheckAgentAlreadyInstalled(kubeClientGetter, log); err != nil {
+		return "", trace.Wrap(err, "could not check if teleport-kube-agent is already installed.")
+	} else if alreadyInstalled {
+		return "", trace.AlreadyExists("teleport-kube-agent is already installed on the cluster %q", clusterName)
```

Comment:
We should give users detailed guides on what to do if chart is installed but pod isn't joinig - expired token, chart installed version with bug, helm chart values drift from the expected ones...

In the future we might want to allow users to upgrade the chart as well

Reply:
Yep, will handle that in the UI. We have a similar thing for when we timeout on waiting for kube agent after installation.